pkg/endpoint/endpoint.go

```diff
@@ -65,44 +66,107 @@ func Listen(ctx context.Context, config Config) (ETCDConfig, error) {
 		return ETCDConfig{}, errors.Wrap(err, "starting kine backend")
 	}
 
-	listen := config.Listener
-	if listen == "" {
-		listen = KineSocket
-	}
-
-	b := server.New(backend)
+	// set up GRPC server and register services
+	b := server.New(backend, endpointScheme(config))
 	grpcServer, err := grpcServer(config)
 	if err != nil {
-		return ETCDConfig{}, err
+		return ETCDConfig{}, errors.Wrap(err, "creating GRPC server")
 	}
-
 	b.Register(grpcServer)
 
-	listener, err := createListener(listen)
+	// set up HTTP server with basic mux
+	httpServer := httpServer()
+
+	// Create raw listener and wrap in cmux for protocol switching
+	listener, err := createListener(config)
 	if err != nil {
-		return ETCDConfig{}, err
+		return ETCDConfig{}, errors.Wrap(err, "creating listener")
+	}
+	m := cmux.New(listener)
+
+	if config.ServerTLSConfig.CertFile != "" && config.ServerTLSConfig.KeyFile != "" {
+		// If using TLS, wrap handler in GRPC/HTTP switching handler and serve TLS
+		httpServer.Handler = grpcHandlerFunc(grpcServer, httpServer.Handler)
+		anyl := m.Match(cmux.Any())
+		go func() {
+			if err := httpServer.ServeTLS(anyl, config.ServerTLSConfig.CertFile, config.ServerTLSConfig.KeyFile); err != nil {
+				logrus.Errorf("Kine TLS server shutdown: %v", err)
+			}
+		}()
+	} else {
+		// If using plaintext, use cmux matching for GRPC/HTTP switching
+		grpcl := m.Match(cmux.HTTP2())
+		go func() {
+			if err := grpcServer.Serve(grpcl); err != nil {
+				logrus.Errorf("Kine GRPC server shutdown: %v", err)
+			}
+		}()
+		httpl := m.Match(cmux.HTTP1())
+		go func() {
+			if err := httpServer.Serve(httpl); err != nil {
+				logrus.Errorf("Kine HTTP server shutdown: %v", err)
+			}
+		}()
 	}
 
 	go func() {
-		if err := grpcServer.Serve(listener); err != nil {
-			logrus.Errorf("Kine server shutdown: %v", err)
+		if err := m.Serve(); err != nil {
+			logrus.Errorf("Kine listener shutdown: %v", err)
+			grpcServer.Stop()
 		}
-		<-ctx.Done()
-		grpcServer.Stop()
-		listener.Close()
 	}()
 
+	endpoint := endpointURL(config, listener)
+	logrus.Infof("Kine available at %s", endpoint)
+
 	return ETCDConfig{
 		LeaderElect: leaderelect,
-		Endpoints:   []string{listen},
+		Endpoints:   []string{endpoint},
 		TLSConfig:   tls.Config{},
 	}, nil
 }
 
-func createListener(listen string) (ret net.Listener, rerr error) {
-	network, address := networkAndAddress(listen)
+// endpointURL returns a URI string suitable for use as a local etcd endpoint.
+// For TCP sockets, it is assumed that the port can be reached via the loopback address.
+func endpointURL(config Config, listener net.Listener) string {
+	scheme := endpointScheme(config)
+	address := listener.Addr().String()
+	if scheme != "unix" {
+		address = "127.0.0.1" + address[strings.LastIndex(address, ":"):]
```

Comment:
Consider leveraging https://pkg.go.dev/net#SplitHostPort

Reply:
thanks! I was looking for something a little more lightweight than a full url parser but didn't come across this.